vllm/model_executor/models/idefics2.py

```diff
@@ -0,0 +1,566 @@
+        self.norm = RMSNorm(self.hidden_size, eps=self.rms_norm_eps)
+
+        self._use_flash_attention_2 =\
+             config._attn_implementation == "flash_attention_2"
+
+    def forward(
+        self,
+        context: torch.Tensor,
+        attention_mask,
+    ) -> torch.Tensor:
+        # seq embed -> bsz seq embed
+        latents = self.latents.unsqueeze(0).expand(
+            (context.shape[0], *self.latents.size()))
+
+        latent_attention_mask = torch.ones(
+            (attention_mask.size(0), latents.size(1)),
+            dtype=attention_mask.dtype,
+            device=attention_mask.device)
+        attention_mask = torch.cat([attention_mask, latent_attention_mask],
+                                   dim=-1)
+        attention_mask = (_prepare_4d_attention_mask(
+            attention_mask, latents.dtype, tgt_len=self.n_latents) if
+                          not self._use_flash_attention_2 else attention_mask)
+
+        compressed_context = latents
+        for perceiver_layer in self.layers:
+            layer_outputs = perceiver_layer(
+                compressed_context,
+                context,
+                attention_mask=attention_mask,
+                position_ids=None,
+                past_key_value=None,
+                output_attentions=False,
+                use_cache=False,
+            )
+
+            compressed_context = layer_outputs[0]
+
+        compressed_context = self.norm(compressed_context)
+
+        return compressed_context
+
+
+class Idefics2Connector(nn.Module):
+
+    def __init__(self, config):
+        super().__init__()
+        self.modality_projection = Idefics2MLP(
+            hidden_size=config.vision_config.hidden_size,
+            intermediate_size=config.text_config.intermediate_size,
+            output_size=config.text_config.hidden_size,
+            hidden_act=config.text_config.hidden_act,
+        )
+        self.perceiver_resampler = Idefics2PerceiverResampler(config)
+
+    def forward(self, image_hidden_states, attention_mask):
+        image_hidden_states = self.modality_projection(image_hidden_states)
+        image_hidden_states = self.perceiver_resampler(
+            context=image_hidden_states, attention_mask=attention_mask)
+        return image_hidden_states
+
+
+class Idefics2Model(nn.Module):
+
+    def __init__(self,
+                 config: "Idefics2Config",
+                 vision_language_config: VisionLanguageConfig,
+                 linear_method: Optional["LinearMethodBase"] = None) -> None:
+        super().__init__()
+        self.config = config
+        self.vision_language_config = vision_language_config
+        self.linear_method = linear_method
+        self.padding_idx = self.config.text_config.pad_token_id
+        self.vocab_size = self.config.text_config.vocab_size
+
+        self.config = config
+        ##SIGLIP vision encodder
+        self.vision_model = SiglipVisionModel(config.vision_config)
```

Comment:
I would leave a note that this is directly using `transformers`'s implementation and it's currently not tensor parallelized.